pkg/controllers/trace_controller_test.go

```diff
@@ -0,0 +1,208 @@
+// Delete() has been called
+func DeleteMethodHasBeenCalled(fakeGadget *FakeGadget, name string) func() bool {
+	key := "delete/" + name
+	return func() bool {
+		f := methodHasBeenCalled(fakeGadget, key)
+		return f() == key
+	}
+}
+
+// UpdatedTrace returns a function that fetches the Trace object in a way that
+// can be used in Gomega's 'Eventually' or 'Consistently' methods.
+func UpdatedTrace(ctx context.Context, key client.ObjectKey) func() *gadgetv1alpha1.Trace {
+	trace := &gadgetv1alpha1.Trace{}
+
+	return func() *gadgetv1alpha1.Trace {
+		err := k8sClient.Get(ctx, key, trace)
+		if err != nil {
+			return nil
+		} else {
+			return trace
+		}
+	}
+}
+
+// HaveState returns a GomegaMatcher that checks if the Trace.Status.State has
+// the expected value
+func HaveState(expectedState string) GomegaMatcher {
+	return WithTransform(func(trace *gadgetv1alpha1.Trace) string {
+		return trace.Status.State
+	}, Equal(expectedState))
+}
+
+// HaveOperationError returns a GomegaMatcher that checks if the
+// Trace.Status.OperationError has the expected value
+func HaveOperationError(expectedOperationError string) GomegaMatcher {
+	return WithTransform(func(trace *gadgetv1alpha1.Trace) string {
+		return trace.Status.OperationError
+	}, Equal(expectedOperationError))
+}
+
+// HaveOutput returns a GomegaMatcher that checks if the Trace.Status.Output
+// has the expected value
+func HaveOutput(expectedOutput string) GomegaMatcher {
+	return WithTransform(func(trace *gadgetv1alpha1.Trace) string {
+		return trace.Status.Output
+	}, Equal(expectedOutput))
+}
+
+// Tests
+
+var _ = Context("Controller with a fake gadget", func() {
+	ctx := context.TODO()
+	gadgetRegistry := make(map[string]gadgets.Gadget)
+	fakeGadget := &FakeGadget{
+		calls: make(map[string]struct{}),
+	}
+	gadgetRegistry["fakegadget"] = fakeGadget
+
+	ns := SetupTest(ctx, gadgetRegistry)
+
+	Describe("when no existing resources exist", func() {
+		It("should create a new Trace resource", func() {
+			traceObjectKey := client.ObjectKey{
+				Name:      "mytrace",
+				Namespace: ns.Name,
+			}
+
+			myTrace := &gadgetv1alpha1.Trace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      traceObjectKey.Name,
+					Namespace: traceObjectKey.Namespace,
+					Annotations: map[string]string{
+						"gadget.kinvolk.io/operation": "magic",
+					},
+				},
+				Spec: gadgetv1alpha1.TraceSpec{
+					Node:       "fake-node",
+					Gadget:     "fakegadget",
+					RunMode:    "Manual",
+					OutputMode: "Status",
+				},
+			}
+
+			err := k8sClient.Create(ctx, myTrace)
+			Expect(err).NotTo(HaveOccurred(), "failed to create test Trace resource")
+
+			Eventually(OperationMethodHasBeenCalled(fakeGadget, traceObjectKey.String(), "magic", "")).Should(BeTrue())
+
+			Eventually(UpdatedTrace(ctx, traceObjectKey)).Should(SatisfyAll(
+				HaveState("FakeState"),
+				HaveOperationError("FakeError"),
+				HaveOutput("FakeOutput"),
```

Comment:
Should it also check that the annotation was removed?

Reply:
Done